### PR TITLE
turn on -Werror=type-limits in our Bazel CPU build

### PR DESCRIPTION
**Pull request by dagitses**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #79159
* #79158
* #79155
* #79156
* #79154
* __->__ #79139

Summary:
We also fix any existing issues. Note that we only do this for the CPU
build because nvcc is considered a C++ toolchain but it does not have
the same flag support. Adding flags to the GPU build will cause nvcc
errors.

Test Plan: Built locally, rely on CI to confirm.

Reviewers: malfet

Subscribers:

Tasks:

Tags:

### CODE REVIEW

**Review by seemethere — APPROVED** [review]

**Review by osalpekar — APPROVED** [review]

**Review by albanD — APPROVED** [review]

File: aten/src/ATen/native/cpu/UnaryOpsKernel.cpp [hunk 1]
```diff
@@ -284,7 +285,7 @@ void sign_kernel(TensorIteratorBase& iter){
 
         cpu_kernel_vec(
           iter,
-          [=](scalar_t a) -> scalar_t { return (0 < a) - (a < 0); },
+          [=](scalar_t a) -> scalar_t { return (0 < a) - c10::is_negative(a); },
```
[the remark this answers is not in the document]
Comment (dagitses) [reply aten/src/ATen/native/cpu/UnaryOpsKernel.cpp:288]: Nope. Because `a` could be equal to zero or positive.

**Pull request closed without merging**